File: src/uk/co/datamantechnology/enigma/EnigmaMachine.java
package uk.co.datamantechnology.enigma;

import java.util.*;

public class EnigmaMachine {

	private int numberOfRotors = 3; // e.g. 3 = this a three rotor enigma
									// machine!
	int wheelOrder[]; // assuming a 3 rotor set up, from left to right we have
						// 2,1,0
	private Reflector reflector;
	private Rotor rotors[];
	private Stecker stecker;

	public EnigmaMachine() {
		setUp();
	}
	
	private void setUp() {
		this.rotors = new Rotor[getNumberOfRotors()];
		this.wheelOrder = new int[getNumberOfRotors()];
		this.reflector = new Reflector(1); // choosing the B reflector;
		this.stecker = new Stecker();
	}
	
	public void setRingSettings(String ringSettings){
		for (int i=0; i < ringSettings.length(); i++){
			this.setRingSetting(ringSettings.length()-i-1,ringSettings.charAt(i));
		}		
	}
	
	public void setRingSetting(int rotorNumber, char letter){
		this.rotors[rotorNumber].setRingSetting(letter);
	}
	
	public void setSteckerPairings(String pairings){
		this.stecker.setPairings(pairings);
	}
	
	public void setWheelOrder(int rotor1, int rotor2, int rotor3) {
		// TODO: refactor to be able to deal with more any number of rotors
		selectRotor(2, rotor1);
		selectRotor(1, rotor2);
		selectRotor(0, rotor3);
	}
	
	private void selectRotor(int wheelOrderPosition, int rotorNumber) {
		wheelOrder[wheelOrderPosition] = rotorNumber - 1;
		this.rotors[wheelOrderPosition] = new Rotor(rotorNumber);
	}

	public void moveRotor(int rotor, String direction) {
		if (direction == "up") {
			moveRotorUp(rotor);
		} else {
			moveRotorDown(rotor);
		}
	}

	public void setIndicators(String indicators) {
		for (int i = 0; i < getNumberOfRotors(); i++) {
			setIndicator(getNumberOfRotors() - i - 1, indicators.charAt(i));
		}
	}

	public String getIndicators(){
		String returnString = "";
		for (int i = getNumberOfRotors()-1; i >= 0 ; i--) {
			returnString+= getIndicator(i);
		}
		return returnString;
	}
	
	public String encrypt(String plainText){
		System.out.println("Encrypting: " + plainText);
		StringBuilder encypheredText = new StringBuilder();
		for (int i=0; i<plainText.length(); i++){
			encypheredText.append(encrypt(plainText.charAt(i)));			
		}
		System.out.println("Encyphered text : " + encypheredText.toString());
		return encypheredText.toString();
	}

	public char encrypt(char clearText) {
		char encypheredChar = clearText;
		
		// go through the stecker		
		encypheredChar = this.stecker.encrypt(clearText);
		
		// move the rotors on one
		advanceRotors();
		// go through the rotors in acsending order (encypher)
		for (int i = 0; i < getNumberOfRotors(); i++) {
			encypheredChar = this.rotors[i].encrypt(encypheredChar);
		}
		// go through the reflector
		encypheredChar = reflector.reflect(encypheredChar);

		// go back through the rotors in descending order (decypher)
		for (int i = getNumberOfRotors() - 1; i >= 0; i--) {
			encypheredChar = this.rotors[i].decrypt(encypheredChar);
		}
		
		// go back trhough the Stecker
		encypheredChar = this.stecker.encrypt(encypheredChar);

		System.out.println(clearText + " " + encypheredChar + " " + getIndicators());
		// return encrypted character
		return encypheredChar;
	}

	private void advanceRotors() {
		List<Integer> rotorsToAdvance = new ArrayList<Integer>();
		// check for notch positions
		for (int i = 0; i < getNumberOfRotors(); i++) {
			// right hand rotor moves on one postion every time;
			if (i == 0){
				rotorsToAdvance.add(i);
			// if preceding rotor is at notch position then this rotor moves on
			}else if (this.rotors[i - 1].isAtNotchPosition()){
				rotorsToAdvance.add(i);
			} else if (this.rotors[i].isAtNotchPosition() && i < (getNumberOfRotors()-1)){
				rotorsToAdvance.add(i);	
			}
		}
		for (int i = 0; i < rotorsToAdvance.size(); i++) {
			this.rotors[rotorsToAdvance.get(i)].advance();
		}
	}

	private void moveRotorUp(int rotor) {
		rotors[rotor].advance();
	}

	private void moveRotorDown(int rotor) {
		rotors[rotor].goBack();
	}

	private void setIndicator(int rotor, char letter) {
		this.rotors[rotor].setIndicator(letter);		
	}

	private char getIndicator(int rotor) {
		return this.rotors[rotor].getIndicator();
	}
	
	
	public void setNumberOfRotors(int numberOfRotors) {
		this.numberOfRotors = numberOfRotors;
	}

	public int getNumberOfRotors() {
		return numberOfRotors;
	}
}

File: python/src/enigmamachine.py
import rotor
import reflector
import stecker

class EnigmaMachine():

    def __init__(self):
        self.rotors = [1, 2, 3]
        self.reflector = reflector.Reflector(1) # choosing the B reflector
        self.stecker = stecker.Stecker()

    def set_ring_settings(self, ringSettings):
        for i in range(len(ringSettings)):
            self.set_ring_setting(len(ringSettings) - i - 1, ringSettings[i])

    def set_ring_setting(self, rotor_number, letter):
        self.rotors[rotor_number].set_ring_setting(letter)

    def set_stecker_pairings(self, pairings):
        self.stecker.set_pairings(pairings)

    def set_wheel_order(self, rotor1, rotor2, rotor3):
        # TODO: refactor to be able to deal with any number of rotors
        self.select_rotor(2, rotor1)
        self.select_rotor(1, rotor2)
        self.select_rotor(0, rotor3)

    def select_rotor(self, wheel_position, rotor_number):
        self.rotors[wheel_position] = rotor.Rotor(rotor_number)

    def moveRotor(self, rotor, direction):
        if (direction == "up"):
            self.moveRotorUp(rotor)
        else:
            self.moveRotorDown(rotor)

    def setIndicators(self, indicators):
        for i in range(self.getNumberOfRotors()):
            self.setIndicator(self.getNumberOfRotors() - i - 1, indicators[i]);

    def getIndicators(self):
        returnString = "";
        print("Number of rotors is: " + str(self.getNumberOfRotors()) + "\r")
        for i in range(self.getNumberOfRotors() - 1, -1, -1):
            returnString += self.getIndicator(i)
        print("Indicators :" + returnString + "\r")
        return returnString;

    def encrypt(self, plainText):
        print ("Encrypting: " + plainText)
        encypheredText = ""
        for i in range(len(plainText)):
            encypheredText += self.encrypt_character(plainText[i])
        print ("Encyphered text : " + encypheredText)
        return encypheredText

    def encrypt_character(self, clearText) :
        encypheredChar = clearText
        print("ENCRYPTING: " + clearText + "\r=============")
        # go through the stecker
        encypheredChar = self.stecker.encrypt(clearText)
        print(encypheredChar + " gone through stecker\r")
        # move the rotors on one
        self.advanceRotors()
        # go through the rotors in acsending order (encypher)
        for i in range(self.getNumberOfRotors()):
            encypheredChar = self.rotors[i].encipher(encypheredChar)
            print(encypheredChar + " gone through rotor" + str(i) + "\r")
        # go through the reflector
        encypheredChar = self.reflector.reflect(encypheredChar)
        print(encypheredChar + " gone through reflector\r")


        # go back through the rotors in descending order (decypher)
        for i in range(self.getNumberOfRotors() - 1, -1, -1):
            encypheredChar = self.rotors[i].decipher(encypheredChar)
            print(encypheredChar + " gone back through rotor" + str(i) + "\r")

        # go back trhough the Stecker
        encypheredChar = self.stecker.encrypt(encypheredChar)
        print(encypheredChar + " back out the stekcer\r")

        # System.out.println()
        # System.out.println(clearText + " " + encypheredChar + " " + getIndicators())
        # return encrypted character
        return encypheredChar

    def advanceRotors(self):
        rotorsToAdvance = []
        # check for notch positions
        for i in range(self.getNumberOfRotors() - 1, -1, -1):
            # right hand rotor moves on one postion every time
            if (i == 0):
                rotorsToAdvance.append(i)
            # if preceding rotor is at notch position then this rotor moves on
            elif (self.rotors[i - 1].isAtNotchPosition()):
                rotorsToAdvance.append(i)
            elif self.rotors[i].isAtNotchPosition() and i < (self.getNumberOfRotors() - 1):
                rotorsToAdvance.append(i)
        for i in range(len(rotorsToAdvance)):
            self.rotors[rotorsToAdvance[i]].advance()

    def moveRotorUp(self, rotor):
        self.rotors[rotor].advance()

    def moveRotorDown(self, rotor):
        self.rotors[rotor].goBack()

    def setIndicator(self, rotor, letter):
        self.rotors[rotor].setIndicator(letter)

    def getIndicator(self, rotor):
        return self.rotors[rotor].getIndicator()


    def setNumberOfRotors(self, numberOfRotors):
        self.numberOfRotors = numberOfRotors

    def getNumberOfRotors(self):
        return len(self.rotors)
